keps/sig-node/4742-node-labels-downward/README.md

```diff
@@ -0,0 +1,812 @@
+The initial design includes:
+
+In KEP 1659, the following labels are defined:
+* topology.kubernetes.io/region
+* topology.kubernetes.io/zone
+
+In addition to the above labels, KEP 1659 declares the entire `topology.kubernetes.io` prefix space as reserved for use by the Kubernetes project.
+
+This KEP expands upon KEP 1659 in the following ways:
+- The `x.topology.kubernetes.io` prefix is allocated for use by end users. The kubernetes project itself will not define any standard labels with that prefix.
+- The `<domain>.x.topology.kubernetes.io` prefix is likewise allocated for use by end users or third-parties. The `<domain>` portion is treated the same as a "normal" label prefix. For example, `example.com.x.topology.k8s.io/label-name`.
+- All labels using the `topology.kubernetes.io` or `*.topology.kubernetes.io` prefix spaces are considered "safe" for workloads. A workload may be exposed to the values of these labels which directly apply to the workload. For example, a pod may learn the topology of the node on which it is running.
+
+The idea is that we will expose those labels from nodes to pods via a literal copy from the Node, for instance using the method `GetNode` from Kubelet in the `podFieldSelectorRuntimeValue` function and `volume.VolumeHost` `GetNodeLabels` function in the `CollectData` function in the downward API.
+
+### User Stories
+
+* As a cluster operator, I want to make decisions based on node topology labels.
+* As a cluster operator, I want to access node topology labels inside of my pod
+* As a cluster operator, I want to access node instance types labels inside of my pod
+* As a developer, I want to know which region my app is serving, to be able to diagnose problems they may face in certain AZs or regions
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+* Scope creep. Allowing additional node information or node label info could
+create security issues. This is mitigated by limiting the node information
+available to the pod to strictly node labels, and specifically those of the
+form `topology.k8s.io/*`.
+
+* Exposing sensitive data as node labels to pods. This is mitigated by ensuring
+node labels contain the specific pattern `topology.k8s.io/*` in order to be
+available to Pods.
+
+* Stale data. Information obtained through node labels is like information
+attained through a configmap or secret mounted to a pod, being passed on
+creation but not guaranteed to be immutable and thus should be treated as so.
+
+
+## Design Details
+
+The intention behind this design is not to add extra structs or fields
+but to use existing logic to extend the current fieldRef as it is used today
+to pass pod metadata to pods using the downwardAPI.
+
+## Implementation Details
```

Comment:
is there any race conditions between pod binding and some controller assigning labels? Anything we should do about it?

Reply:
For example, if it is implemented as an init container with permissions, there may be implementation that waits for those labels. If it's natively implemented by kubelet, there is no synchronization possible and people switching from one model to another may get into the race condition situation